Linker problem when installing alex on CentOS 5 with newer cabal-install

Okay, yikes, esoteric issue title, I know.

On a 32-bit CentOS 5 machine, I've got three installations of `cabal-install` and `Cabal`:

```
$ /opt/ghc-7.6.3/bin/cabal --version
cabal-install version 1.18.0.3
using version 1.18.1.3 of the Cabal library
$ /opt/ghc-7.8.3/bin/cabal --version
cabal-install version 1.20.0.3
using version 1.20.0.1 of the Cabal library
$ cabal --version
cabal-install version 1.22.0.0
using version 1.22.0.0 of the Cabal library
```

Keeping GHC versions the same at 7.8.3, just switching `cabal`s, I try installing `alex` into a fresh sandbox. The 1.18.0.3 install goes swimmingly. The 1.20.0.3 and 1.22.0.0 installs however fail to link the `alex` executable at the end of the build:

```
...
Linking dist/dist-sandbox-79015d48/build/alex/alex ...
/tmp/tmp.ISEzb16554/.cabal-sandbox/lib/i386-linux-ghc-7.8.3/tf-random-0.5/libHStf-random-0.5.a(Gen.o):(.text+0x3bb): undefined reference to `Threefish_256_Process_Block'
/tmp/tmp.ISEzb16554/.cabal-sandbox/lib/i386-linux-ghc-7.8.3/tf-random-0.5/libHStf-random-0.5.a(Gen.o):(.text+0x543): undefined reference to `Threefish_256_Process_Block'
/tmp/tmp.ISEzb16554/.cabal-sandbox/lib/i386-linux-ghc-7.8.3/tf-random-0.5/libHStf-random-0.5.a(Gen.o):(.text+0x883): undefined reference to `Threefish_256_Process_Block'
/tmp/tmp.ISEzb16554/.cabal-sandbox/lib/i386-linux-ghc-7.8.3/tf-random-0.5/libHStf-random-0.5.a(Gen.o):(.text+0xb9f): undefined reference to `Threefish_256_Process_Block'
/tmp/tmp.ISEzb16554/.cabal-sandbox/lib/i386-linux-ghc-7.8.3/tf-random-0.5/libHStf-random-0.5.a(Gen.o):(.text+0xd5b): undefined reference to `Threefish_256_Process_Block'
/tmp/tmp.ISEzb16554/.cabal-sandbox/lib/i386-linux-ghc-7.8.3/tf-random-0.5/libHStf-random-0.5.a(Gen.o):(.text+0xee3): more undefined references to `Threefish_256_Process_Block' follow
collect2: ld returned 1 exit status
Failed to install alex-3.1.4
Updating world file...
cabal: Error: some packages failed to install:
alex-3.1.4 failed during the building phase. The exception was:
ExitFailure 1
```

I notice that the final invocation of `ghc` to build the executable has more linker flags in the newer versions. Maybe something is causing the linker to miss the symbol from `tf-random`'s cbits?

I've posted full logs here: https://gist.github.com/acfoltzer/c46274098956d03e6f79

It's worth noting that this does not occur on a similarly-configured 64-bit machine, nor any of our newer CentOS machines.

Thanks for any ideas you might have.

--- a/Cabal/Distribution/Simple/Program/Builtin.hs
+++ b/Cabal/Distribution/Simple/Program/Builtin.hs
@@ -264,7 +264,24 @@ arProgram :: Program
 arProgram = simpleProgram "ar"
 
 stripProgram :: Program
-stripProgram = simpleProgram "strip"
+stripProgram = (simpleProgram "strip") {
+    programFindVersion =
+      findProgramVersion "--version" $ \str ->
+        -- Invoking "strip --version" gives very inconsistent
+        -- results. We look for the first word that starts with a
+        -- number, and try parsing out the first two components of
+        -- it. Non-GNU 'strip' doesn't appear to have a version flag.
+        let numeric ""    = False
+            numeric (x:_) = isDigit x
+        in case dropWhile (not . numeric) (words str) of
+          (ver:_) ->
+            -- take the first two version components
+            let isDot         = (== '.')
+                (major, rest) = break isDot ver
+                minor         = takeWhile (not . isDot) (dropWhile isDot rest)
+            in major ++ "." ++ minor
+          _ -> ""
+  }
 
 hsc2hsProgram :: Program
 hsc2hsProgram = (simpleProgram "hsc2hs") {

--- a/Cabal/Distribution/Simple/Program/Strip.hs
+++ b/Cabal/Distribution/Simple/Program/Strip.hs
@@ -11,10 +11,13 @@ module Distribution.Simple.Program.Strip (stripLib, stripExe)
        where
 
 import Distribution.Simple.Program (ProgramConfiguration, lookupProgram
-                                   ,rawSystemProgram, stripProgram)
+                                   , programVersion, rawSystemProgram
+                                   , stripProgram)
 import Distribution.Simple.Utils   (warn)
-import Distribution.System         (Platform(..), OS (..), buildOS)
+import Distribution.System         (Arch(..), Platform(..), OS (..), buildOS)
+import Distribution.Text           (simpleParse)
 import Distribution.Verbosity      (Verbosity)
+import Distribution.Version        (withinRange)
 
 import Control.Monad               (unless)
 import System.FilePath             (takeBaseName)
@@ -42,7 +45,7 @@ stripExe verbosity (Platform _arch os) conf path =
        _   -> []
 
 stripLib :: Verbosity -> Platform -> ProgramConfiguration -> FilePath -> IO ()
-stripLib verbosity (Platform _arch os) conf path = do
+stripLib verbosity (Platform arch os) conf path = do
   case os of
     OSX -> -- '--strip-unneeded' is not supported on OS X, iOS or
            -- Solaris. See #1630.
@@ -53,6 +56,17 @@ stripLib verbosity (Platform _arch os) conf path = do
                -- libraries with lots identically named modules. See
                -- #1784.
                return()
+    Linux | arch == I386 ->
+      -- Versions of 'strip' on 32-bit Linux older than 2.18 are
+      -- broken. See #2339.
+      let (Just okVersion) = simpleParse ">= 2.18"
+      in case programVersion =<< lookupProgram stripProgram conf of
+          Just v | withinRange v okVersion ->
+            runStrip verbosity conf path args
+          _ -> warn verbosity $ "Unable to strip library '"
+                                ++ (takeBaseName path)
+                                ++ "' (version of 'strip' too old; "
+                                ++ "requires >= 2.18 on 32-bit Linux)"
     _   -> runStrip verbosity conf path args
   where
     args = ["--strip-unneeded"]

